storagemarket/network/deal_stream_v110.go

```diff
@@ -0,0 +1,82 @@
+package network
+
+import (
+	"bufio"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/mux"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	cborutil "github.com/filecoin-project/go-cbor-util"
+
+	"github.com/filecoin-project/go-fil-markets/storagemarket/migrations"
+)
+
+type dealStreamv110 struct {
+	p        peer.ID
+	host     host.Host
+	rw       mux.MuxedStream
+	buffered *bufio.Reader
+}
+
+var _ StorageDealStream = (*dealStreamv110)(nil)
+
+func (d *dealStreamv110) ReadDealProposal() (Proposal, cid.Cid, error) {
+	var ds migrations.Proposal1
+
+	if err := ds.UnmarshalCBOR(d.buffered); err != nil {
+		err = fmt.Errorf("unmarshalling v110 deal proposal: %w", err)
+		log.Warnf(err.Error())
+		return ProposalUndefined, cid.Undef, err
+	}
+
+	proposalNd, err := cborutil.AsIpld(ds.DealProposal)
+	if err != nil {
+		err = fmt.Errorf("getting v110 deal proposal as IPLD: %w", err)
+		log.Warnf(err.Error())
+		return ProposalUndefined, cid.Undef, err
+	}
+
+	prop, err := migrations.MigrateClientDealProposal0To1(*ds.DealProposal)
+	if err != nil {
+		err = fmt.Errorf("migrating v110 deal proposal to current version: %w", err)
+		log.Warnf(err.Error())
+		return ProposalUndefined, cid.Undef, err
+	}
```

Comment:
Is this from the user? We can't do this migration without (possibly) changing signatures.

Reply:
Good catch 👍 

One option would be to reject non-utf8 labels sent over version 1.1.0 of the deal proposal protocol.

Reply:
After discussing with @magik6k and @arajasek I implemented the proposed change: reject deal proposals over the v1.1.0 protocol where the deal label is not parseable utf8

Reply:
@dirkmc just for my sanity, this will only affect any new deals made, not historical deals that may have non-utf8 labels, yes?

Reply:
Historical deals will be migrated from the old format (deal label as a string) to the new format (DealLabel struct).
If those deals were made with a non-utf8 deal label, their signature will no longer match the new format. It's my understanding that this doesn't matter because those deals have already been published on chain.

This PR introduces a new protocol version. So we now have:
- v1.1.0: The current version
  When a deal proposal is received, reject the deal proposal if it has a label that cannot be parsed as utf8.
  Otherwise, migrate the deal proposal to the new format (label as DealLabel struct). This should not change the signature of the deal because the label is CBOR-marshalled in the same was as if it were a string. See the tests in [storagemarket/network/deal_stream_v110_test.go](https://github.com/filecoin-project/go-fil-markets/pull/721/files#diff-300884c916912bfde5ed40e33b85b32560a88e8cb3af07d4dea76be1ebdfe20d)
- v1.1.1: The new version
  This protocol version will be used when both the client and provider use a version of lotus that depends on this PR.
  Deals made over v1.1.1 are already in the new format (label as a DealLabel struct).

Reply:
@dirkmc Got it, I think this is good for any newly made / received deals. My concern is whether the entire migration of historical deals will fail if there are historical non-utf8 deals (which I agree will no longer have valid signatures) -- and if so, what is the impact of this migration failing?

(also when does the migration of historical deal info happen? at startup?)

Reply:
The migration happens at startup 👍 

The migration is not expected to fail for historical deals with non-utf8 labels. Please take a look at the code here:
https://github.com/filecoin-project/go-fil-markets/pull/721/files#diff-8536cd1664598e5c06e27807ae8fa95a9f1d40439d4883fbef351e54826dbeccR248
If there's an issue with that code let me know.

Reply:
Perfect, thanks, that assuages my concern -- I think I was conflating the `ReadDealProposal` (which uses `checkDealLabel` and fails if non-utf-8) with the migration code, that succeeds in an immaterially wrong state.

Thanks for indulging!

Reply:
Ah yep, that makes sense 👍